cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -237,7 +242,48 @@ protected void shutDown() throws Exception {
         result.add(enforceApplicationDetailAccess(entry.getKey(), applicationDetail));
       }
     }
-    return result;
+   return result;
+  }
+
+  public void scanApplications(NamespaceId namespace, Set<String> artifactNames,
+      @Nullable String artifactVersion,
+      Consumer<ApplicationDetail> consumer) {
+
+    Predicate<ApplicationDetail> predicate = getAppPredicate(artifactNames, artifactVersion);
+
+    store.scanApplications(20,
+        (appId, appSpec) -> {
+          if (!appId.getNamespace().equals(namespace.getNamespace())) {
+            return;
+          }
+
+          Set<? extends EntityId> visible = accessEnforcer.isVisible(ImmutableSet.of(appId),
+              authenticationContext.getPrincipal());
+          if (!visible.contains(appId)) {
+            return;
+          }
+
+          Map<ApplicationId, String> owners;
+          try {
+            owners = ownerAdmin.getOwnerPrincipals(ImmutableSet.of(appId));
```

Comment:
Same as above, this call will query the owner store, which translate to a SQL query. This can be expensive if you do it one app at a time.